src/main/java/com/minecolonies/coremod/entity/ai/citizen/herders/AbstractEntityAIHerder.java

```diff
@@ -0,0 +1,534 @@
+     */
+    public final List<ToolType>  toolsNeeded = new ArrayList<>();
+    public final List<ItemStack> itemsNeeded = new ArrayList<>();
+
+    /**
+     * Amount of animals needed to bread.
+     */
+    public static final int NUM_OF_ANIMALS_TO_BREED = 2;
+
+    /**
+     * Butchering attack damage.
+     */
+    private static final int BUTCHERING_ATTACK_DAMAGE = 5;
+
+    /**
+     * Distance two animals need to be inside to breed.
+     */
+    private static final int DISTANCE_TO_BREED = 10;
+
+    /**
+     * Amount of animals to keep per Hut Level.
+     */
+    private final int maxAnimalMultiplier;
+
+    /**
+     * Delays used to setDelay()
+     */
+    public static final int DELAY_TWENTY      = 20;
+    public static final int DELAY_FOURTY      = 40;
+    public static final int DELAY_ONE_HUNDRED = 100;
+
+    /**
+     * Number of actions needed to dump inventory.
+     */
+    private static final int ACTIONS_FOR_DUMP = 10;
+
+    /**
+     * Creates the abstract part of the AI.
+     * Always use this constructor!
+     *
+     * @param job the job to fulfill.
+     * @param maxAnimalsMultiplier the number of max animals to add per hut level
+     */
+    public AbstractEntityAIHerder(@NotNull final J job, final int maxAnimalsMultiplier)
+    {
+        super(job);
+        this.maxAnimalMultiplier = maxAnimalsMultiplier;
+        super.registerTargets(
+          new AITarget(IDLE, START_WORKING),
+          new AITarget(START_WORKING, this::startWorkingAtOwnBuilding),
+          new AITarget(PREPARING, this::prepareForHerding),
+          new AITarget(HERDER_DECIDE, this::decideWhatToDo),
+          new AITarget(HERDER_BREED, this::breedAnimals),
+          new AITarget(HERDER_BUTCHER, this::butcherAnimals),
+          new AITarget(HERDER_PICKUP, this::pickupItems)
+        );
+        worker.setCanPickUpLoot(true);
+    }
+
+    @Override
+    protected int getActionsDoneUntilDumping()
+    {
+        return ACTIONS_FOR_DUMP;
+    }
+
+    @NotNull
+    @Override
+    protected List<ItemStack> itemsNiceToHave()
+    {
+        final List<ItemStack> list = super.itemsNiceToHave();
+        list.add(new ItemStack(getBreedingItem(), 2));
+        return list;
+    }
+
+    /**
+     * Decides what job the herder should switch to, breeding or Butchering.
+     *
+     * @return The next {@link AIState} the herder should switch to, after executing this method.
+     */
+    public AIState decideWhatToDo()
+    {
+        setDelay(DELAY_FOURTY);
+
+        final List<T> animals = new ArrayList<>(searchForAnimals());
+
+        if (animals.isEmpty())
+        {
+            setDelay(DELAY_ONE_HUNDRED);
+            return HERDER_DECIDE;
+        }
+
+        worker.setLatestStatus(new TextComponentTranslation("com.minecolonies.coremod.status.herder.deciding"));
+
+        final int numOfBreedableAnimals = animals.stream().filter(animal -> animal.getGrowingAge() == 0).toArray().length;
```

Comment:
there are count/sum methods on streams, converting to an array adds a lot of unnececary overhead

Reply:
Didn't notice that one sorry. different names XD